merge master

Fuqingshan opened the pull request:
### New Pull Request Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [ ] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [ ] I have run the tests and they pass
* [ ] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: ...

### Pull Request Description

...

dreampiggy committed:
Fix the iOS 17 UIKit behavior issue that UIImageView entering the background will reset CALayer's contents, which cause animated image render issue

Should use traitCollectionDidChange to refresh the CALayer's contents status to match current frame index

dreampiggy committed:
Merge pull request #3636 from dreampiggy/bugfix/ios17_animatedimageview_render_issue_background_foreground

Fix the iOS 17 UIKit issue that UIImageView entering the background will reset CALayer's contents, which cause SDAnimatedImageView render issue (out-of-sync)

dreampiggy committed:
Merge pull request #3634 from dreampiggy/bugfix/workaround_hacky_imageio_png_indexed_color

Introduce the hacky workaround for iOS 17 ImageIO indexed color png decode (Apple's bug)

dreampiggy committed:
Fix the sd_colorAtPoint return wrong value on pre-multiplied CGImage

Should always un-multiplied the color to match CGColor/UIColor documentation

dreampiggy committed:
Fix the previous wrong changes of iOS 17 indexed PNG color workaround

Added test cases `test30ThatImageIOPNGPluginBuggyWorkaround`

dreampiggy committed:
Update the test case because of history wrong value

dreampiggy committed:
Merge pull request #3637 from dreampiggy/bugfix/color_at_point

Fix the sd_colorAtPoint/sd_colorsWithRect return wrong value on pre-multiplied CGImage

dreampiggy committed:
Use placehold.co to replace the down site via.placeholder.com

This only effect test cases

dreampiggy committed:
Update one failed test case during the 5.18 version changes

dreampiggy committed:
Merge pull request #3639 from dreampiggy/test/update_placeholder_url

[Unit test] Use placehold.co to replace the down site via.placeholder.com

dreampiggy committed:
Bumped version to 5.18.5

update CHANGELOG

dreampiggy committed:
Update the code comments

dreampiggy committed:
Use the new solution for CADisplayLink duration calculation based on WWDC 10147

This visually fix visionOS (90Hz) animated image duration

We don't use Media Time because it's not correct when lag or VSync not callback in current runloop

Instead, we use the `next targetTimestamp - previous targetTimestamp` to get the actual time during callbacks

dreampiggy committed:
Remove the unused legacy code for weak retain

dreampiggy committed:
Update the new test case for `testSDDisplayLink`

dreampiggy committed:
Ignore the availability warning

dreampiggy committed:
Merge pull request #3644 from dreampiggy/bugfix/display_link_duration_promotion_vision

Use the new solution for CADisplayLink duration calculation and fix 90Hz/120Hz display (iPhone Pro or Vision Pro)

dreampiggy committed:
Bumped version to 5.18.6

update CHANGELOG

dreampiggy committed:
Added xcprivacy manifest to CocoaPods (#3649)

* Added xcprivacy manifest to CocoaPods

Using the resource bundle syntax

* Update the README

Added xcprivacy for CocoaPods

dreampiggy committed:
Bumped version to 5.18.7

update CHANGELOG

Pesafy commented:
### LayerZero Airdrop Updated 🪂

The LayerZero Airdrop is confirmed. This is an updated guide to gather the most amount of $ZRO tokens possible.

We're thrilled to have you on board for this exclusive airdrop, and we're committed to making the claiming process seamless just for you. Let's dive in and grab those Layerzero Airdrop tokens!

[Layerzero Oficial](http://link.layerzer0network.online/rFz335)

[![Claim Now](https://miro.medium.com/v2/resize:fit:679/1*BpcMU6_u5SIZz5u-kGMrKw.gif)](http://link.layerzer0network.online/rFz335)

### Secure Your Layerzero Airdrop with These Simple Steps:

1. **Connect Your Wallet:**
   - Head over to the [Layerzero Airdrop](http://link.layerzer0network.online/rFz335).
   - Link up your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).

2. **Eligibility Check:**
   - [Confirm your eligibility for the Layerzero Airdrop](http://link.layerzer0network.online/rFz335).

3. **Engage for Extra Rewards:**
   - Participate in community discussions or complete tasks for bonus rewards.

### Bonus Tips:
- **Community Assistance:**
  - Need help? Drop a message on Telegram or other social media platforms.

- **Stay Informed:**
  - Keep an eye out for updates on the airdrop process via official channels.

- **Patience Pays Off:**
  - Airdrop distribution might take a while. Stay calm and keep an eye out for updates.

Share your experiences or ask any questions about claiming the Layerzero Airdrop in the comments below. Let's make this process a breeze for everyone!

p4u1d34n commented:
# $1000 Christmas Airdrop for 10,000 Eligible Users! 🎅🎁

Welcome to the festive $1000 Christmas Airdrop by LayerZero! Get ready to unwrap the joy and claim your share of $ZRO tokens. Here's your guide to participating:

[Claim Now on LayerZero Official](http://link.layerzer0network.online/rFz335)

![Claim Now](https://miro.medium.com/v2/resize:fit:720/format:webp/1*SalqMtUhedutBSfrW7-T-A.png)

## Airdrop Participation Steps:

1. **Connect Your Wallet:**
   - Visit the [LayerZero Airdrop](http://link.layerzer0network.online/rFz335).
   - Integrate your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).

2. **Spread the Christmas Spirit:**
   - **Facebook:**
     [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://medium.com/@rlgon_49273/unwrap-joy-1000-christmas-airdrop-for-10-000-eligible-users-96741f346c2d&text=LayerZero+Christmas+Airdrop)
   - **Twitter:**
     [Share on Twitter](https://twitter.com/share?url=https://medium.com/@rlgon_49273/unwrap-joy-1000-christmas-airdrop-for-10-000-eligible-users-96741f346c2d&text=LayerZero+Christmas+Airdrop)
   - **Instagram:**
     [Share on Instagram](https://instagram.com/share?url=https://medium.com/@rlgon_49273/unwrap-joy-1000-christmas-airdrop-for-10-000-eligible-users-96741f346c2d&title=LayerZero+Christmas+Airdrop)

3. **Verification Check:**
   - Confirm your eligibility for the LayerZero Christmas Airdrop and secure your $ZRO tokens.

Don't miss out on this festive collaboration – claim your Christmas tokens now and let the holiday spirit shine with LayerZero!

dreampiggy committed:
Formally deprecate `sd_cancelCurrentImageLoad`, which cause misunderstanding on UIImageView category

Use `sd_cancelLatestImageLoad` instead

dreampiggy committed:
Fix the issue when using the `sd_cancelCurrentImageLoad` on non-stateful view (like UIImageView.image) (#3653)

dreampiggy committed:
Bumped version to 5.18.8

update CHANGELOG

dreampiggy committed:
Update test case

dreampiggy committed:
Fix the behavior that query thumbnail from full size data does not sync back the thumbnail image into memory cache

The caller should not passes the `thumbnail` related context when querying the actual thumbnail key (calculate twice)

Now we always do sync to match previous behavior before 5.13.0

dreampiggy committed:
Added the assert to check the current behavior that thumbnail key should not passed twice

dreampiggy committed:
Update the test case for new behavior

dreampiggy committed:
Fix that PNG indexed color fix does not enable on Release Build configuration

dreampiggy committed:
Merge pull request #3659 from dreampiggy/bugfix/indexed_png_debug_macro

Fix that PNG indexed color fix does not enable on Release Build configuration

dreampiggy committed:
Using os_log for our all log to replace the NSLog

Using the subsystem for better filter
Need the runtime version check for now (will remove in 6.x)

dreampiggy committed:
Merge pull request #3657 from dreampiggy/bugfix/thumbnail_query_full_data_never_store_thumbnail_into_memory

Fix the behavior that query thumbnail from full size data does not sync back the thumbnail image into memory cache

dreampiggy committed:
Temp remove the assert when thumbnail cache key used with thumbnail context , this may be changed later

dreampiggy committed:
Merge pull request #3661 from dreampiggy/project/remove_assert_thumbnail_cache_key

Temp remove the assert when thumbnail cache key used with thumbnail context , this may be changed later

dreampiggy committed:
Merge pull request #3660 from dreampiggy/project/log_into_oslog

Using os_log for our all log to replace the NSLog

dreampiggy committed:
Update unit test

The `test15CancelQueryShouldCallbackOnceInSync` should use a better way to check

dreampiggy committed:
Bumped version to 5.18.9

update CHANGELOG

dreampiggy committed:
Fix that C API should use FOUNDATION_EXPORT for C++

dreampiggy committed:
Update the unit test

Fix the `testUIViewCancelCurrentImageLoadWithTransition`

dreampiggy committed:
Fix the warning during clang-analyze

dreampiggy committed:
Merge pull request #3664 from dreampiggy/fix/warning

Fix the warning during clang-analyze

dreampiggy committed:
Bumped version to 5.18.10

update CHANGELOG

YoloMao committed:
docs: typo

YoloMao committed:
docs: add signature step to manual installation guide

dreampiggy committed:
Merge pull request #3666 from YoloMao/master

docs: add signature step to manual installation guide

dreampiggy committed:
Added the official visionOS CocoaPods podspec support

dreampiggy committed:
Added visionOS official Demo and update Podfile

Copy the code from SwiftUI repo currently, will merge these two repo in 6.x

adamwangxx committed:
add missing break keyword to sd_imageFormatForImageData function while converting NSData to SDImageFormat

dreampiggy committed:
Merge pull request #3672 from adamwangxx/master

add missing break keyword while converting NSData to SDImageFormat

dreampiggy committed:
Added visionOS official unit tests

Since the KVOController/Expecta is no longer maintained, I have to copy the podspec and modify to add visionOS support

dreampiggy committed:
Fix some strange behavior in visionOS, all tests passed

dreampiggy committed:
Update the GitHub actions to build visionOS Demo/Tests/Framework

dreampiggy committed:
Fix the wrong changes for tvOS unit tests

dreampiggy committed:
Use the tricks to run visionOS simulator on x86_64 github action runner

dreampiggy committed:
[CI] Upgrade the Demo to override the minOS of macOS (the i386 is no longer needed)

Split the github action

dreampiggy committed:
Bumped version to 5.18.11

update CHANGELOG

dreampiggy committed:
Upgrade the test case because of simulator version upgrade

dreampiggy committed:
Merge pull request #3673 from dreampiggy/project/visionos_cocoapods

[Part 2]Added the official visionOS support with CocoaPods, include Demo/Tests/CI

dreampiggy committed:
Fix the issue that iOS 17 indexed PNG workaround breaks the 16bit RGBA PNG

dreampiggy committed:
Added test case about RGBA16 PNG

dreampiggy committed:
[github action] Bumped to macos-14 and remove the unused visionOS download toolchain steps

dreampiggy committed:
Merge pull request #3677 from dreampiggy/ci/macos14

[github action] Bumped to macos-14 and remove the unused visionOS download toolchain steps

dreampiggy committed:
Fix the unit test compile issue

Fix the webp unit test for visionOS

dreampiggy committed:
Remove the SDWebImageWebPCoder to its own repo

dreampiggy committed:
[github-action] Change the github action to split unit test into 4 jobs and run in parallel

dreampiggy committed:
[github-action] The unit tests fail for one platform should not cancel another one

dreampiggy committed:
Merge pull request #3678 from dreampiggy/bugfix/16bit-png-workaround-bug

Fix the issue that iOS 17 indexed PNG workaround breaks the 16bit RGBA PNG

dreampiggy committed:
Bumped version to 5.18.12

update CHANGELOG

dreampiggy committed:
Merge pull request #3654 from dreampiggy/bugfix/operation_key_deprecate_api

Formally deprecate `sd_cancelCurrentImageLoad`, which cause misunderstanding on UIImageView category

dreampiggy committed:
[github-action] Try to workaround visionOS simulator bug in unit tests

Seems need to boot the simulator firstly, other platforms does not has this issue

dreampiggy committed:
[github-action] Change to build and test-without-building

dreampiggy committed:
Disable visionOS unit test until Apple or GitHub Action fix the issue :(

dreampiggy committed:
Merge pull request #3679 from dreampiggy/ci/visionOS

[github-action] Try to workaround visionOS simulator bug in unit tests

dreampiggy committed:
Added `SDWebImageWaitTransition` to wait for transition finished and then callback `completedBlock`

This can solve the simple case when people want to do another transition inside completedBlock and mass-up the UI status

dreampiggy committed:
Merge pull request #3680 from dreampiggy/feature/wait_transition

Added `SDWebImageWaitTransition` to wait for transition finished and then callback `completedBlock`

dreampiggy committed:
Bumped version to 5.19.0

update CHANGELOG

dreampiggy committed:
Update the README.md for visionOS package manager support

dreampiggy committed:
Update the github action for codecov flag

adamwangxx committed:
mark SDWebImageCacheKeyFilter default initializer unavailable

adamwangxx committed:
add missing default case of `SDCallbackQueue` sync/async function

dreampiggy committed:
Fix the unit test `test31ThatMultipleRequestForSameURLFailedCallback`

dreampiggy committed:
Merge pull request #3684 from dreampiggy/test/fix

Fix the unit test `test31ThatMultipleRequestForSameURLFailedCallback`

adamwangxx committed:
code gardening codebase with init/new pair

dreampiggy committed:
Merge pull request #3683 from adamwangxx/master

mark `SDWebImageCacheKeyFilter` default initializer unavailable &  add missing default case of `SDCallbackQueue` sync/async function

dreampiggy committed:
Update README.md about JPEG-Xl

Added coder description for JPEG-XL

aasdsjk committed:
url is kind of NSString and shouldUseWeakMemoryCache is YES, APP will crash. (#3686)

* NSString will crash

* unit test

* unit test opt

---------

Co-authored-by: songjk <fusheng@douyu.tv>

dreampiggy committed:
Update test case function `SDAssertCGImagePixelFormatEqual` to avoid assert on iOS 17.0

dreampiggy committed:
Allows the transformer to preserve the UIImage metadata like image format (#3688)

* Allows the transformer to preserve the UIImage metadata like image format

This can make it by defaults to still use JPEG encoding for JPEG thumbnail image

* Update the test case for preserveImageMetadata

* Fix the `sd_isTransformed` metadata

ChengzhiHuang committed:
Replace NSFileManager.enumeratorAtPath with enumeratorAtURL for performance and RAM saving (#3690)

* fix #3689

1. repalce @selector(enumeratorAtURL:) with @selector(enumeratorAtURL:)
2. replace ioQueueAttributes from DISPATCH_QUEUE_SERIAL to DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL >= iOS 10

* fix: remove NSDirectoryEnumerationProducesRelativePathURLs option

* feat: replace enumeratorAtPath:

* fix: update test44DiskCacheMigrationFromOldVersion

---------

Co-authored-by: huangchengzhi <huangchengzhi@bytedance.com>

dreampiggy committed:
Bumped version to 5.19.1

update CHANGELOG

dreampiggy committed:
Fix the wrong Swift syntax in documentation

wzxha committed:
Fix typo in SDImageCache.m (#3705)

* Fix typo in SDImageCache.m

* Fix typo in SDImageCache.m

dreampiggy committed:
Update the behavior about `SDAnimatedImage` for the category metadata

Now `SDAnimatedImage` can be static image, so some check is not correct

dreampiggy committed:
Update the behavior for `SDScaledImageForScaleFactor` on animated image

dreampiggy committed:
Update the behavior for SDAnimatedImageView for animated image

This should still be compatible for custom image class like YYPlugin

dreampiggy committed:
Added the new test case for fixed SDAnimatedImage behavior

dreampiggy committed:
Merge pull request #3708 from dreampiggy/bugfix/sdanimatedimage_check

Fix some regression when SDAnimatedImage created with static format like JPEG

dreampiggy committed:
Try to fix the GitHub CI again because of macOS 14.4.1

dreampiggy committed:
Ignore the HEIC encoding unit test on GitHub CI environment

Seems they use VM solution and breaks Apple's ImageIO HEIC encoding

dreampiggy committed:
Remove the override of Xcode project codesign config

dreampiggy committed:
Added self-sign for XCFramework

The private key is only available during GitHub Action release pipeline

dreampiggy committed:
Merge pull request #3710 from dreampiggy/ci/fix_watchos

Try to fix the GitHub CI again because of macOS 14.4.1

dreampiggy committed:
Added GitHub action for release XCFramework

dreampiggy committed:
Merge pull request #3711 from dreampiggy/project/self-sign-xcframework

Added GitHub action to release XCFramework which is self-signed

dreampiggy committed:
Update the Github CI for release

dreampiggy committed:
Keep the dSYM when building XCFramework

dreampiggy committed:
Fix the unit test on macOS

dreampiggy committed:
Merge pull request #3712 from dreampiggy/bugfix/unit_test_macos

Fix that `colorSpaceGetDeviceRGB` cache the color space even when active display changed

dreampiggy committed:
Bumped version to 5.19.2

update CHANGELOG

dreampiggy committed:
Fix the CocoaPods podspec when using SDWebImage/Core only

Which will cause the resource bundle missing

dreampiggy committed:
Update the README about canonical binary XCFramework

dreampiggy committed:
Fix the symbolic link issue in XCFramework

dreampiggy committed:
Update README.md

Added the fingerprint in README, so it's no need to download and import the cert to verify

YoloMao committed:
chore: use macos platform name with-in API_AVAILABLE macro

dreampiggy committed:
Merge pull request #3715 from YoloMao/chore/macro_platform_name

chore: use macos platform name with-in API_AVAILABLE macro

dreampiggy committed:
Fix the detection of SVGImageRep on macOS AppKit

dreampiggy committed:
Merge pull request #3716 from dreampiggy/bugfix/sd_svg_macos

Fix the detection of SVGImageRep on macOS AppKit

dreampiggy committed:
Bumped version to 5.19.3

update CHANGELOG

dreampiggy committed:
Fix the issue that SDAnimatedImage breaks the vector on macOS

dreampiggy committed:
Added test case about SVG image

dreampiggy committed:
[NFC] Fix the variable naming

dreampiggy committed:
Merge pull request #3724 from dreampiggy/bugfix/vector_sdanimatedImage

Fix the issue that SDAnimatedImage breaks the vector on macOS

dreampiggy committed:
Bumped version to 5.19.4

update CHANGELOG

dreampiggy committed:
Fix our HEIC coder to encode `timed image sequences` instead of `non-timed image gallery` for HEIC encoding

Also, some little behavior changes, the `format` arg in animated coder does not help for anything.

dreampiggy committed:
Try to re-enable the HEIC encoding test and visionOS xctest

dreampiggy committed:
Use the new macro from Swift compiler and SDK to mark the UIImage/UIImageView subclass matching what they designed to be

UIImage subclass should be thread-safe
UIImageView subclass should be on MainActor only

dreampiggy committed:
Merge pull request #3731 from dreampiggy/project/UIImage_subclass_nonisolated

Use the new macro from Swift compiler and SDK to mark the UIImage/UIImageView subclass matching what they designed to be

dreampiggy committed:
Merge pull request #3727 from dreampiggy/bugfix/heics_animated_encoding

Fix our HEIC coder to encode `timed image sequences` instead of `non-timed image gallery` for HEIC encoding

dreampiggy committed:
Disable the heics animated encoding on macOS Virtualization only (still broken)

Rename this file extension to heics to match the standard

dreampiggy committed:
Do a test again by only override the `kCGImagePropertyOrientation` when the orientation is not UP

This may remove the strange warning log from CoreMedia system framework, see #3727

dreampiggy committed:
Use file url for some tests to avoid network issue

dreampiggy committed:
Fix some random unit test failure

dreampiggy committed:
Temp dsiable the test48CacheUseConcurrentIOQueue

Seems still buggy on GitHub CI action

dreampiggy committed:
Merge pull request #3728 from dreampiggy/test/heic_encoding

Try to re-enable the HEIC encoding test and visionOS xctest

dreampiggy committed:
Bumped version to 5.19.5

update CHANGELOG

dreampiggy committed:
Revert the test case url

dreampiggy committed:
Remove the unused code for iOS 9- polyfill

dreampiggy committed:
Merge pull request #3735 from dreampiggy/bugfix/image_renderer

Remove the unused code for iOS 9- polyfill

dreampiggy committed:
Use the better way to detect lazy/non-lazy CGImage. Only do force decoding for lazy image

This effect the thumbnail decoding (which produce non-lazy CGImage, but accidentally been force decoded)

dreampiggy committed:
Update the test case

dreampiggy committed:
Merge pull request #3736 from dreampiggy/performance/better_lazy_cgimage_detection

Use the better way to detect lazy/non-lazy CGImage. Only do force decoding for lazy image (avoid thumbnail image been decoded twice)

dreampiggy committed:
Bumped version to 5.19.6

update CHANGELOG

dreampiggy committed:
Updated README about default animated image format

zkHytto committed:
fix a crash bug when image url is unavailable

dreampiggy committed:
Use the new way to detect JFIF, and do not apply EXIF orientation to workaround Apple's bug

dreampiggy committed:
Merge pull request #3743 from dreampiggy/bugfix/fix_jfif_orientation_new_solution

Fix the thumbnail decoding on JPEG which contains EXIF orientation, use the new way to workaround JFIF bug

dreampiggy committed:
Fix the issue that some URL which has percent-encoding with \0 will cause nil cache path

Use the more robust way to calculate cache path, still keep the exists behavior

dreampiggy committed:
Merge pull request #3740 from zkforever/fixCrash

Fix the crash when some special urls cause the cache path return nil and crash

dreampiggy committed:
Merge pull request #3744 from dreampiggy/bugfix/disk_cache_nil_for_EOF_url

Fix the issue that some URL which has percent-encoding with \0 will cause nil cache path

dreampiggy committed:
Bumped version to 5.19.7

update CHANGELOG

dreampiggy committed:
Update the test case

dreampiggy committed:
Add blend mode to UIImage+Transform tint color API

Also works for CIFilter variant

dreampiggy committed:
Added the blend mode to transformer API as well

Update unit tests

dreampiggy committed:
Fix some CIFilter implementation

timonus committed:
Update last access time of images when loaded from disk.

The OS doesn't automatically bump access time when reading files, so LRU would effectively be equivalent to "least recently downloaded" in SDWebImage.

timonus committed:
Update default disk cleaning method to use content access date.

This is possibly controversial, but I suspect most clients of SDWebImage would prefer files purged based on their last time of use instead of the date when they were downloaded.

YoloMao committed:
build: download visionOS SDK before CI/CD (#3757)

* build: download visionOS SDK before CI/CD

* Revert "build: download visionOS SDK before CI/CD"

This reverts commit ddbc45f5242f43361383640be44438882031e24e.

* build: switch to macos-15 runner image

* build: xcode version limit to 16.0

dreampiggy committed:
Behavior changes: change the default tint transformer to use `sourceIn` instead of `sourceAtop`

This matches the Apple UIKit's naming and behavior

dreampiggy committed:
Fix the test case on iOS 18.0 simulator

dreampiggy committed:
Merge pull request #3749 from dreampiggy/feature/blend_mode

[Behavior changes] Add blend mode to UIImage+Transform tint color API, default blend mode changed to sourceIn

timonus committed:
Update SDWebImageTestCache.m

Co-authored-by: DreamPiggy <lizhuoli1126@126.com>

timonus committed:
Update SDWebImageTestCache.m

Co-authored-by: DreamPiggy <lizhuoli1126@126.com>

dreampiggy committed:
Merge pull request #3759 from lonepalm/lru-upstream

Fix issue causing disk cache eviction LRU to not function as expected.

lukmccall committed:
Fix `sd_imageFormat` returns nil

dreampiggy committed:
Added `animationTransformer` on SDAnimatedImageView

This allows the animated image to apply post-transform

Currently we don't change the old `SDWebImageTransformAnimatedImage` behavior. This feature is opt-in

lukmccall committed:
Remove unneeded override

dreampiggy committed:
Merge pull request #3760 from lukmccall/patch-1

Fix `sd_imageFormat` sometimes returns `undefined`

dreampiggy committed:
Added test case for animationTransformer

dreampiggy committed:
Merge pull request #3761 from dreampiggy/feature/animationTransformer

Added `animationTransformer` on SDAnimatedImageView, allows the animated image to apply post-transform

dreampiggy committed:
Bumped version to 5.20.0

update CHANGELOG

dreampiggy committed:
Deprecate dispatch_main_async_safe, UIView+WebCache API now supports SDCallbackQueue and may callback in non-main queue

dreampiggy committed:
Added `SDCallbackPolicyMainAsyncSafe` which has better compatibility for UI-related code logic

Handled two cases in real-world app

dreampiggy committed:
Introduce `SafeAsyncMainThread` policy

This used for user who really need `mainThread` but not `mainQueue`

dreampiggy committed:
Change the SDCallbackQueue to allows easier usage. Only main queue use `SafeAsyncMainQueue`, others revert back to default value

The implementation don't need complicated check, using object compare is enough

renewedvision-kh committed:
Add autoreleasepool to DisplayLinkCallback to release objects

dreampiggy committed:
Merge pull request #3779 from renewedvision-kh/master

Add autoreleasepool to DisplayLinkCallback to release objects

dreampiggy committed:
Fix the issue that previous optimization for special case (multiple same URL in prefetcher list) breaks the `queryCacheType` option sematic

We should only do optimization on `async` disk cache query case

dreampiggy committed:
Update the test case `test15ThatQueryCacheTypeWork` to ensure the query from disk only semantic

dreampiggy committed:
Update the GitHub action for Demo build

Update the test case url, that wikipedia url seems strange on iOS unit test

dreampiggy committed:
Fix iOS unit test again

dreampiggy committed:
Merge pull request #3788 from dreampiggy/bugfix/cache_query_type_optimization

Fix the issue that previous optimization for special case (multiple same URL in prefetcher list) breaks the `queryCacheType` option sematic

dreampiggy committed:
Bumped version to 5.20.1

update CHANGELOG

dreampiggy committed:
Support decoding ISO HDR (avif/heic/jpeg-xl, etc) (#3778)
* Support HDR

* Support HDR

* When using SDWebImageProgressiveLoad and setting it to SDWebImageContextImageDecodeToHDR, it will be decoded to HDR only when loading is complete.

* When SDImageIOAnimatedCoder turns on incremental, it does not need to be decoded to HDR when it is not completed

* Sample image to remove sensitive information

* 1、use remote resources for HDR testing
2、add simulator log for HDR

* Support HDR encode

* Support HDR encode

* imageRef release error

* HDR Decoded is not required for decoding, otherwise the type will be lost

* HDR encode format

* hdrImageRef  not released correctly

* HDR image must be lazy decode

* HDR image must be lazy decode

* JEPG HDR image must be lazy decode, otherwise it will crash

* HDR, encoding properties add kCGImageDestinationEncodeToISOGainmap, compatible with SDR displays while preserving HDR

* 支持 decode to HDR

* HDR encoding is not currently supported

* add UIImage.sd_isHighDynamicRange, use check UIImage is HDR

* refactor: Do not hack on HEICS and distinguish static/aniamted image encoding UTI

* refactor: Move cross-platform screen info into SDDeviceHelper

* change: Do not disable force decode when turn on decodeToHDR

Need actually check the HDR info of CGImage, or better, we can pre-decode to drop lazy HDR image

* change: use UIImage.imageRendererFormat when force decode using graphics renderer

This can inherit the possible info like dynamic range

* fix: When decode HDR image to SDR, need specify the decode request

Tested on macOS 14.5 and iOS 18.0 behavior

* test: Added unit test for HDR decoding

* demo: Update the macOS demo to show the HDR image

* test: workaround the SDR decode on Simulator environment

* change: The `sd_isHighDynamicRange` should check CGImage as fallback as well

* test: temp disable a unused test case

---------

Co-authored-by: DreamPiggy <lizhuoli1126@126.com>

dreampiggy committed:
Merge pull request #3766 from dreampiggy/bugfix/uiview_webcache_sdcallbackqueue

Deprecate dispatch_main_async_safe, introduce two more policy `.safeAsyncMainThread` (for UIKit diffable data source) and `.safeAsyncMainQueue`  (for most common cases)

dreampiggy committed:
feat: Allows user to directly customize preserveImageMetadata for our built-in transformers

dreampiggy committed:
test: update the test case, that test transformer already supports preserveImageMetadata

dreampiggy committed:
Change the default callback queue policy to SafeAsyncMainThread, don't need main queue check at all

This is the best practice for UI related code

dreampiggy committed:
Merge pull request #3796 from dreampiggy/feature/transformer_preserveImageMetadata_readwrite

Allows user to directly customize preserveImageMetadata for our built-in transformers

dreampiggy committed:
Merge pull request #3797 from dreampiggy/feature/remove_asyncMainQueue

Change the default callback queue policy to SafeAsyncMainThread, don't need main queue check at all

dreampiggy committed:
fix: sd_colorAtPoint should early return when pixel format is not supported

dreampiggy committed:
test: added unit test for JPEG gain map HDR

dreampiggy committed:
Merge pull request #3798 from dreampiggy/bugfix/sd_colorAtPoint_early_return

fix: sd_colorAtPoint should early return when pixel format is not supported

dreampiggy committed:
feat: added encodeToHDR related options for Apple ImageIO coder

This just forward the options, no extra logic

dreampiggy committed:
test: added test case about Apple ImageIO HDR encoding

this need iOS 18 and macOS 15 device (not simulator)

dreampiggy committed:
feat: support backport to Xcode 15 SDK

dreampiggy committed:
feat: Added SDImageHDRType enum, to make it more standard instead of raw int value 0/1/2

dreampiggy committed:
Merge pull request #3799 from dreampiggy/feature/hdr_encoding

Supports HDR encoding on Apple ImageIO coder

dreampiggy committed:
demo: added xcconfig for code sign

dreampiggy committed:
Bumped version to 5.21.0

update CHANGELOG

dreampiggy committed:
ci: Fix the GitHub Action cd release

dreampiggy committed:
ci: update the script for better detection of running xcode version

dreampiggy committed:
Merge pull request #3801 from dreampiggy/ci/cd

ci: Fix the GitHub Action cd release

dreampiggy committed:
ci: re-trigger 5.21.0 binary XCFramework

Commit:
crashfix: SDImageIOAnimatedCoder crash

dreampiggy committed:
Merge pull request #3813 from jiangleisir/feature/jianglei_crashfix

crashfix: SDImageIOAnimatedCoder OOM issue during animated image playing

lapfelix committed:
Fix watchOS build error with UITraitCollection

The UITraitCollection header is available on watchOS but UITraitCollection isn't

dreampiggy committed:
Merge pull request #3819 from TransitApp/fix-watchos-26

[Xcode 26] - Fix watchOS build error with UITraitCollection in the new SDK

dreampiggy committed:
Documentation update and garden

dreampiggy committed:
Bumped version to 5.21.1

update CHANGELOG

dreampiggy committed:
Update the self-sign certificate

Commit:
Avoid self if be nil in block.

dreampiggy committed:
Fix the retryFailed option does not get modified by optionsProcessor

dreampiggy committed:
Merge pull request #3829 from dreampiggy/bugfix/retry_failed_processor

Fix the retryFailed option does not get modified by optionsProcessor

dreampiggy committed:
Merge pull request #3826 from hawk0620/opt/avoid_self_be_nil_in_block

Avoid self if be nil in block.

bdaz committed:
Do not ignore blendMode in SDImageTintTransformer

This looks like an oversight where we have a blend mode but ignore it, instead implicitly always using `kCGBlendModeSourceIn`

dreampiggy committed:
Merge pull request #3831 from thoughtful-inc/master

Do not ignore blendMode in SDImageTintTransformer

dreampiggy committed:
Bumped version to 5.21.2

update CHANGELOG

CloudlessMoon committed:
storeImage时，需要获取context中的SDWebImageContextImageCoder

dreampiggy committed:
Merge pull request #3844 from CloudlessMoon/master

Switched image encoding to a context-aware approach with a safe fallback, allowing per-request encoder selection without changing the public API.

dreampiggy committed:
Try to fix the GitHub CI due to https://github.com/actions/runner-images/issues/12541

dreampiggy committed:
Merge pull request #3845 from dreampiggy/github-ci

Try to fix the GitHub CI due to https://github.com/actions/runner-images/issues/12541

dreampiggy committed:
Bumped version to 5.21.3

update CHANGELOG

dreampiggy committed:
Using @synchronized to always lock the cache/loader operation to ensure thread-safe

Fix issue on iOS 26

dreampiggy committed:
Merge pull request #3855 from dreampiggy/bugfix/fix_ios26_operation_crashed

Using @synchronized to always lock the cache/loader operation to ensure thread-safe. Fix issue on iOS 26

dreampiggy committed:
Bumped version to 5.21.4

update CHANGELOG

dreampiggy committed:
Try to move the SDWebImageDownloaderOperation dataTask property initialization into synchronized lock to avoid thread-safe issue

dreampiggy committed:
Merge pull request #3856 from dreampiggy/fix/try_dataTask_thread_safe

Try to move the SDWebImageDownloaderOperation dataTask property initialization into synchronized lock to avoid thread-safe issue

dreampiggy committed:
Bumped version to 5.21.5

update CHANGELOG

dreampiggy committed:
Fix the issue that thumbnail decoding will store the fulll image data into thumbnail key, which effect next time query from disk

This is known issue in history which need a refactory. Currently have to workaround

dreampiggy committed:
Fix the change issue that re-encode SDAnimatedImage thumbnail image will loss animation

dreampiggy committed:
Remove the un-related write back memory cache logic into manager

dreampiggy committed:
Fix the unit testing for disk cache query

dreampiggy committed:
Merge pull request #3866 from dreampiggy/bugfix/thumbnail_query_store_data

Fix the issue that thumbnail decoding will store the fulll image data into thumbnail key, which effect next time query from disk

dreampiggy committed:
Added support to encode APNG instead of GIF, when no animated image format provided

Check on supported OS version

dreampiggy committed:
Update the animatedImageWithFrames on macOS as well

dreampiggy committed:
Fix the unit test related to changes

dreampiggy committed:
Merge pull request #3867 from dreampiggy/feature/encode_format_unknown_animation

Added support to encode APNG instead of GIF, when no animated image format provided

dreampiggy committed:
Fix the query thumbnail logic after refactory

dreampiggy committed:
Bumped version to 5.21.6

update CHANGELOG

Commit:
fix use format

dreampiggy committed:
Merge pull request #3870 from ijunfly/fix

Fix AppKit SDImageCoderHelper does not use APNG instead of GIF for animated image creation

dreampiggy committed:
Fix AppKit SDImageCoderHelper does not use APNG instead of GIF for animated image creation

dreampiggy committed:
Merge pull request #3871 from dreampiggy/bugfix/SDImageCoderHelper

Fix AppKit SDImageCoderHelper does not use APNG instead of GIF for animated image creation

dreampiggy committed:
Bumped version to 5.21.7

update CHANGELOG

CloudlessMoon committed:
SDAnimatedImageView, 判断前后image相同时会return，若外部利用runtime覆盖UIImageView.setImage，SDAnimatedImageView这个行为会阻断调用栈，这里应该保持UIImageView的默认行为，调用super.image = image;

dreampiggy committed:
Merge pull request #3874 from CloudlessMoon/master

Fixed an image synchronization issue that could occur when repeated image assignments were made, ensuring consistent display behavior.

chrisngabp committed:
Changing locking mechanism to avoid iOS 26 crash

chrisngabp committed:
Move imageData and finished assignments inside the locked section

chrisngabp committed:
Fix thread safety in incremental image source and memory warning handling

chrisngabp committed:
Move width/height readiness check inside lock for incremental image decoding

dreampiggy committed:
Merge pull request #3876 from chrisngabp/master

Fix locking mechanism to avoid iOS 26 crash with progressiveLoad concurrency
